receiver/cudareceiver/collector.go

```diff
@@ -0,0 +1,179 @@
+	device, status := NVMLDeviceGetHandledByIndex(uint64(0))
+	if status != NVMLSuccess {
+		return nil, fmt.Errorf("Could not get GPU device: status=%d", status)
+	}
+
+	c := &CUDAMetricsCollector{
+		consumer:       consumer,
+		startTime:      time.Now(),
+		device:         device,
+		scrapeInterval: d,
+		metricPrefix:   prefix,
+		done:           make(chan struct{}),
+	}
+
+	return c, nil
+}
+
+func detectResource() {
+	resourceDetectionSync.Do(func() {
+		res, err := auto.Detect(context.Background())
+		if err != nil {
+			panic(fmt.Sprintf("Resource detection failed, err:%v", err))
+		}
+		if res != nil {
+			rsc = &resourcepb.Resource{
+				Type:   res.Type,
+				Labels: make(map[string]string, len(res.Labels)),
+			}
+			for k, v := range res.Labels {
+				rsc.Labels[k] = v
+			}
+		}
+	})
+}
+
+// StartCollection starts a ticker'd goroutine that will scrape and export CUDA metrics periodically.
+func (c *CUDAMetricsCollector) StartCollection() {
+	detectResource()
+	status := NVMLInit()
+	if status != NVMLSuccess {
+		// TODO: handle inappropriate GPU state
+		return
+	}
+
+	go func() {
+		ticker := time.NewTicker(c.scrapeInterval)
+		for {
+			select {
+			case <-ticker.C:
+				c.scrapeAndExport()
+
+			case <-c.done:
+				return
+			}
+		}
+	}()
+}
+
+// StopCollection stops the collection of metric information
+func (c *CUDAMetricsCollector) StopCollection() {
+	status := NVMLShutdown()
+	if status != NVMLSuccess {
+		// TODO: handle inappropriate GPU state
+		return
+	}
+	close(c.done)
+}
+
+// scrapeAndExport
+func (c *CUDAMetricsCollector) scrapeAndExport() {
+	ctx := context.Background()
+
+	metrics := make([]*metricspb.Metric, 0, len(cudaMetricDescriptors))
+	var errs []error
+
+	metrics = append(
+		metrics,
+		&metricspb.Metric{
+			MetricDescriptor: metricTemperature,
+			Resource:         rsc,
+			Timeseries:       []*metricspb.TimeSeries{c.getInt64TimeSeries(c.device.Temperature())},
+		},
+		&metricspb.Metric{
+			MetricDescriptor: metricPower,
+			Resource:         rsc,
+			Timeseries:       []*metricspb.TimeSeries{c.getInt64TimeSeries(c.device.PowerUsage())},
+		},
+		&metricspb.Metric{
+			MetricDescriptor: metricPCIeThroughputTX,
+			Resource:         rsc,
+			Timeseries:       []*metricspb.TimeSeries{c.getInt64TimeSeries(c.device.PCIeThroughput(PCIeUtilTXBytes))},
+		},
+		&metricspb.Metric{
+			MetricDescriptor: metricPCIeThroughputRX,
+			Resource:         rsc,
+			Timeseries:       []*metricspb.TimeSeries{c.getInt64TimeSeries(c.device.PCIeThroughput(PCIeUtilRXBytes))},
+		},
+	)
+
+	if len(errs) > 0 {
+		// TODO: emit error log
+		return
+	}
+
+	c.consumer.ConsumeMetricsData(ctx, consumerdata.MetricsData{Metrics: metrics})
+}
+
+// TimeToTimestamp converts a time.Time to a timestamp.Timestamp pointer.
+// TODO: remove this function once it gets exposed.
+// https://github.com/open-telemetry/opentelemetry-collector/blob/master/internal/internal.go
```

Comment:
Hmm I'm not sure why this internal function exists. Can't we just use the official function?

https://github.com/golang/protobuf/blob/master/ptypes/timestamp.go#L59

Reply:
I use the official function here. I just followed the implementation in the upstream.
https://github.com/open-telemetry/opentelemetry-collector/blob/master/receiver/vmmetricsreceiver/vm_metrics_collector.go#L225

Reply:
Yeah - in my opinion upstream is wrong :) But you're also not using the official function here, you're copying it. Can't you call `ptypes.TimestampProto(t time.Time)` instead of reimplementing this again?

Reply:
ah, I meant I used the function "in the coming commit". Sorry for confusion.